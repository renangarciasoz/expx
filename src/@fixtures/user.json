{
  "username": "test user",
  "name": "Test User",
  "email": "user@expx.fi",
  "contracts": [
    {
      "id": "0o301",
      "symbol": "eth_lido",
      "holding": 325.1
    }
  ]
}
