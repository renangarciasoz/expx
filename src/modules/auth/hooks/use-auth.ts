import { api } from "api/api-client";
import {
  recoverUserInformation,
  signInRequest,
  SignInRequestData,
} from "api/auth.api";
import { AxiosError } from "axios";
import { EXPIRATION_IN_HOURS, TOKEN_NAME } from "constants/auth.constants";
import { DASHBOARD, HOME } from "constants/urls.constants";
import { useRouter } from "next/router";
import { destroyCookie, parseCookies, setCookie } from "nookies";
import { useEffect } from "react";
import { useStore } from "store/auth.store";

export function useAuth() {
  const router = useRouter();
  const { user, error, setUser, setError, removeUser, ...store } = useStore();
  const isAuthenticated = !!user;

  useEffect(() => {
    const { [TOKEN_NAME]: token } = parseCookies();

    if (token) {
      recoverUserInformation().then((user) => {
        setUser(user);
      });
    }
  }, [setUser]);

  async function signIn({ email, password }: SignInRequestData) {
    setError("");

    try {
      const { token, ...user } = await signInRequest({
        email,
        password,
      });

      setCookie(undefined, TOKEN_NAME, token, {
        maxAge: 60 * 60 * EXPIRATION_IN_HOURS, // seconds * minutes * constant number in hours
      });

      api.defaults.headers.common["Authorization"] = `Bearer ${token}`;

      setUser(user);

      router.push(DASHBOARD);
    } catch (e) {
      removeUser();
      delete api.defaults.headers.common["Authorization"];
      setError((e as AxiosError).message);
    }
  }

  async function signOut() {
    destroyCookie(undefined, TOKEN_NAME);
    delete api.defaults.headers.common["Authorization"];
    removeUser();
    router.push(HOME);
  }

  return {
    ...store,
    isAuthenticated,
    user,
    error,
    signIn,
    setError,
    signOut,
  };
}
