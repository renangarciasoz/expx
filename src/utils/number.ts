import { ONE_BILLION } from "constants/numbers.constants";

export const formatValue = (value: number, minimumFractionDigits = 2) =>
  Intl.NumberFormat("en", {
    style: "currency",
    currency: "USD",
    minimumFractionDigits,
  }).format(value);

export const formatBillionValue = (
  value: string | number,
  minimumFractionDigits = 3
) => {
  return `${formatValue(Number(value) / ONE_BILLION, minimumFractionDigits)}b`;
};
