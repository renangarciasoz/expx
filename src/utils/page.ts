export const blockPageScroll = () => {
  const body = document?.querySelector("body");
  body!.style.height = "100vh";
  body!.style.overflow = "hidden";
};

export const unblockPageScroll = () => {
  const body = document?.querySelector("body");
  body!.style.height = "auto";
  body!.style.overflow = "auto";
};
