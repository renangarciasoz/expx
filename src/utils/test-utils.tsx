import { render, RenderOptions } from "@testing-library/react"; //jest-ignore
import { i18n } from "configs/i18next-tests.config";
import { ThemeProvider } from "contexts/theme.ctx";
import React, { ReactElement } from "react";
import { I18nextProvider } from "react-i18next";

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <ThemeProvider>
      <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
    </ThemeProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
