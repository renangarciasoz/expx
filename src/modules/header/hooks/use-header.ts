import { POSITION_TOP } from "constants/header.constants";
import { useStore as useAuthStore } from "store/auth.store";
import { useStore } from "store/header.store";
import { blockPageScroll, unblockPageScroll } from "utils/page";

export const useHeader = () => {
  const { position, setFormAuthIsOpened, ...store } = useStore();
  const { setError } = useAuthStore();

  const headerIsAtTopPosition = position === POSITION_TOP;

  function openAuthForm() {
    blockPageScroll();
    setFormAuthIsOpened(true);
  }

  function closeAuthForm() {
    unblockPageScroll();
    setFormAuthIsOpened(false);
    setError(null);
  }

  return {
    openAuthForm,
    closeAuthForm,
    headerIsAtTopPosition,
    ...store,
  };
};
