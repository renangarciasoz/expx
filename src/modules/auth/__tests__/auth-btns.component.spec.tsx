/**
 * @jest-environment jsdom
 */

import user from "@fixtures/user.json";
import * as useAuth from "modules/auth/hooks/use-auth";
import * as useHeader from "modules/header/hooks/use-header";
import { fireEvent, render, screen } from "utils/test-utils";
import { AuthBtns } from "../auth-btns.component";

const useAuthSpy = jest.spyOn(useAuth, "useAuth");
() => ({
  ...jest.requireActual("modules/auth/hooks/use-auth"),
});

const useHeaderSpy = jest.spyOn(useHeader, "useHeader");
() => ({
  ...jest.requireActual("modules/heade/hooks/use-header"),
});

describe("Auth Btns", () => {
  describe("when renders buttons", () => {
    it("renders correctly", () => {
      const { container } = render(<AuthBtns />);
      expect(container).toMatchSnapshot();
    });

    describe("when user is not authenticated", () => {
      beforeEach(() => {
        render(<AuthBtns />);
      });

      it("renders login button", () => {
        expect(screen.getByTestId("login-btn")).toBeInTheDocument();
      });

      describe("user clicks on login button", () => {
        beforeEach(() => {
          fireEvent.click(screen.getByTestId("login-btn"));
        });

        it("renders close login button", () => {
          expect(screen.getByTestId("close-login-btn")).toBeInTheDocument();
        });
      });
    });

    describe("when user is authenticated", () => {
      beforeEach(() => {
        useAuthSpy.mockImplementation(() => ({
          ...jest.requireActual("modules/auth/hooks/use-auth"),
          isAuthenticated: true,
          user,
        }));

        useHeaderSpy.mockImplementation(() => ({
          ...jest.requireActual("modules/header/hooks/use-header"),
          authFormIsOpened: false,
        }));

        render(<AuthBtns />);
      });

      it("renders logout button and user name", () => {
        const logoutButton = screen.getByTestId("logout-btn");
        const username = screen.getByText(user.username);

        expect(logoutButton).toBeInTheDocument();
        expect(username).toBeInTheDocument();
      });
    });
  });
});
