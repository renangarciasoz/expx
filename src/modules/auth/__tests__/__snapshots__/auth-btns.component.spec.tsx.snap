// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Btns when renders buttons renders correctly 1`] = `
<div>
  <div
    class="MuiBox-root css-1j9bsaa"
  >
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButtonBase-root  css-xn5ipf-MuiButtonBase-root-MuiButton-root"
      data-testid="login-btn"
      tabindex="0"
      type="button"
    >
      Login
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
